src/Parser.php

```diff
@@ -33,39 +33,104 @@ class Parser implements \IteratorAggregate, PositionAware
     private $lexer;
 
     /** @var array */
-    private $jsonPointerPath;
+    private $jsonPointerPaths;
 
-    /** @var string */
+    /** @var array */
     private $jsonPointer;
 
+    /** @var string */
+    private $currentJsonPointer;
+
     /** @var Decoder */
     private $jsonDecoder;
 
     /**
      * @param Traversable $lexer
-     * @param string $jsonPointer Follows json pointer RFC https://tools.ietf.org/html/rfc6901
-     * @param Decoder $jsonDecoder
+     * @param array|string $jsonPointer Follows json pointer RFC https://tools.ietf.org/html/rfc6901
+     * @param Decoder|null $jsonDecoder
+     * @throws InvalidArgumentException
      */
     public function __construct(Traversable $lexer, $jsonPointer = '', $jsonDecoder = null)
     {
-        if (0 === preg_match('_^(/(([^/~])|(~[01]))*)*$_', $jsonPointer, $matches)) {
-            throw new InvalidArgumentException(
-                "Given value '$jsonPointer' of \$jsonPointer is not valid JSON Pointer"
-            );
-        }
-
         $this->lexer = $lexer;
-        $this->jsonPointer = $jsonPointer;
-        $this->jsonPointerPath = array_slice(array_map(function ($jsonPointerPart){
-            return str_replace(
-                '~0', '~', str_replace('~1', '/', $jsonPointerPart)
-            );
-        }, explode('/', $jsonPointer)), 1);
         $this->jsonDecoder = $jsonDecoder ?: new ExtJsonDecoder(true);
+        $this->buildJsonPointerPaths((array)$jsonPointer);
+    }
+
+    /**
+     * @param array $jsonPointers
+     * @throws InvalidArgumentException
+     */
+    private function buildJsonPointerPaths(array $jsonPointers)
+    {
+        $jsonPointers = array_values($jsonPointers);
+
+        foreach ($jsonPointers as $jsonPointerEl) {
+            if (preg_match('_^(/(([^/~])|(~[01]))*)*$_', $jsonPointerEl) === 0) {
+                throw new InvalidArgumentException(sprintf("Given value '%s' of \$jsonPointer is not valid JSON Pointer", $jsonPointerEl));
+            }
+
+            $intersectingJsonPointers = array_filter($jsonPointers, static function($el) use ($jsonPointerEl) {
```

Comment:
What if the pointers with wildcard intersect? Let's say `['/list/-/items', '/list/1/items', '/list/1/items/item']`?

Reply:
I've amended the code to also check for wildcards